rust-runtime/inlineable/src/client_request_compression.rs

```diff
@@ -0,0 +1,256 @@
+            options: Some(options),
+        });
+        cfg.push_layer(layer);
+
+        Ok(())
+    }
+
+    fn modify_before_signing(
+        &self,
+        context: &mut BeforeTransmitInterceptorContextMut<'_>,
+        _runtime_components: &RuntimeComponents,
+        cfg: &mut ConfigBag,
+    ) -> Result<(), BoxError> {
+        let state = cfg
+            .load::<RequestCompressionInterceptorState>()
+            .expect("set in `read_before_serialization`");
+
+        let options = state.options.clone().unwrap();
+        let request = context.request_mut();
+
+        // Don't wrap a body if compression is disabled.
+        if !options.is_enabled() {
+            tracing::trace!("request compression is disabled and will not be applied");
+            return Ok(());
+        }
+
+        // Don't wrap a body if it's below the minimum size
+        //
+        // Because compressing small amounts of data can actually increase its size,
+        // we check to see if the data is big enough to make compression worthwhile.
+        if let Some(known_size) = http_body::Body::size_hint(request.body()).exact() {
+            if known_size < options.min_compression_size_bytes() as u64 {
+                tracing::trace!("request body is below minimum size and will not be compressed");
+                return Ok(());
+            }
+            tracing::trace!("compressing non-streaming request body...")
+        } else {
+            tracing::trace!("compressing streaming request body...");
+        }
+
+        wrap_request_body_in_compressed_body(
+            request,
+            CompressionAlgorithm::Gzip.into_impl_http_body_0_4_x(&options),
+        )?;
+
+        Ok(())
+    }
+}
+
+fn wrap_request_body_in_compressed_body(
+    request: &mut HttpRequest,
+    request_compress_impl: Box<dyn CompressRequest>,
+) -> Result<(), BuildError> {
+    request.headers_mut().append(
+        request_compress_impl.header_name(),
+        request_compress_impl.header_value(),
+    );
+    let mut body = {
+        let body = mem::replace(request.body_mut(), SdkBody::taken());
+        body.map(move |body| {
+            let body = CompressedBody::new(body, request_compress_impl.clone());
+
+            SdkBody::from_body_0_4(body)
+        })
+    };
+    mem::swap(request.body_mut(), &mut body);
+
+    Ok(())
+}
+
+#[derive(Debug, Copy, Clone, Default)]
+pub(crate) struct DisableRequestCompression(pub(crate) bool);
+
+impl From<bool> for DisableRequestCompression {
+    fn from(value: bool) -> Self {
+        DisableRequestCompression(value)
+    }
+}
+
+impl Storable for DisableRequestCompression {
+    type Storer = StoreReplace<Self>;
+}
+
+#[derive(Debug, Copy, Clone)]
+pub(crate) struct RequestMinCompressionSizeBytes(pub(crate) u32);
+
+impl Default for RequestMinCompressionSizeBytes {
+    fn default() -> Self {
+        RequestMinCompressionSizeBytes(10240)
+    }
+}
+
+impl From<u32> for RequestMinCompressionSizeBytes {
```

Comment:
A trait implementation exposes a type in `pub` automatically. I see `RequestMinCompressionSizeBytes` stores `u32` as `pub(crate)`, which keeps `u32` private, but the `impl From` sort of undoes it, exposing `u32` to the world. If we can get way with a `pub(crate)` conversion, I think we should prefer that ([reference](https://youtu.be/qfknfCsICUM?si=qHwnvYZoEOh2BBnN&t=2780)). If not, let's stick to what we have here (we can add another `impl From<xxx>` in the future in an additive fashion.

The same goes for the above `impl From<bool> for DisableRequestCompression`.